### ISSUE TITLE
Filenames/Foldernames are url encoded

### ISSUE THREAD

**Issue by alex-LE** [s3tools/s3cmd#554]
I've recently updated to the latest master version, but now s3cmd reupload a lot of already existing files but as a url encoded name like:

already existing: "/folder foo bar"
new created folder: "/folder+foo+bar"

the same happens to the files inside those folders.

Any ideas how to fix it?

**Comment by fviard** [comment]
Hi,
What server are your using? official AWS S3 servers?

**Comment by alex-LE** [comment]
No, dreamobjects by dreamhost

Am Donnerstag, 18. Juni 2015 schrieb Florent Viard :

**Comment by fviard** [comment]
Ok, and how are you looking at the file? through their web interface?
Because if they don't do urldecode before displaying in their interface, it is a bug you should report to them.
Maybe @mdomsch can give you some hint as I think that he his using dreamhost.

**Comment by alex-LE** [comment]
I'm using Cyberduck on a Mac.
In the log i see the names with spaces, but with Cyberduck i actually see a
new folder with + instead of space (the folder with spaces already exists).

2015-07-02 20:32 GMT+02:00 Florent Viard <notifications@github.com>:

**Comment by alex-LE** [comment]
I've started a new upload with a new bucket (the same source files) and now all files/folders names are "wrong" (with the plus sign). Is seems that just spaces are being replaced, umlauts like "ü" are not being replaced.
I can send you a screenshot via email, if you want.

Alex

**Comment by mustafa-muhammad** [comment]
I am using the master branch and still get this error, every space ' ' is turned into '+', using debug I get:

DEBUG: DeUnicodising u'test2 2' using UTF-8
DEBUG: DeUnicodising u'test2 2' using UTF-8
DEBUG: DeUnicodising u'test2 2' using UTF-8
DEBUG: String 'test2 2' encoded to 'test2+2'
DEBUG: CreateRequest: resource[uri]=/test2+2
Any idea how to deal with this?

**Comment by mustafa-muhammad** [comment]
After replacing quote_plus with quote in /S3/S3.py this works.

**Comment by fviard** [comment]
@mustafa-muhammad @alex-LE So it is a bug in Cyberduck that is not doing the proper "decode". "Quote_plus" is what is expected to by amazon S3.
(http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html)

If you remove the quote_plus, you will encounter random issues with some filenames.

**Comment by mustafa-muhammad** [comment]
Actually I use ceph rgw, what issues you think will happen?

**Comment by fviard** [comment]
That is the standard, so first, if you use anything else than cyberduck to restore your files, the filenames will be scrambled. Like "+" will become spaces and so.
I guess that you will encounter "SignatureDoesNotMatch" error from time to time.

**Comment by mustafa-muhammad** [comment]
Thank you, one more thing (not relevent here, sorry), when I copy from bucket to another, when I list the destination bucket with --list-md5, the copied files don't have an MD5, is this by design?

**Comment by mustafa-muhammad** [comment]
So " " and "+" are encoded as "+"? How to retrieve the original file name? I tested with "C++ Guide" and got "C+++Guide", how it is supposed to differentiate?

**Comment by fviard** [comment]
@mustafa-muhammad : No it is not like that, but only the way it is decoded/displayed in your app.
There is 2 choices for encoding:
- "quote" : that encode " "(space) to %20 and "+" to %2B
- "quote_plus": that encode " " to + and "+" to %2B.
And at the opposite there is unquote and unquote_plus.
If you use unquote with a string that was quoted with "quote_plus". The "+" will stay "+" and the "%2B" will become a "+".

```
>>> from urllib import quote, quote_plus
>>> quote("Hello world+foo")
'Hello%20world%2Bfoo'
>>> quote_plus("Hello world+foo")
'Hello+world%2Bfoo'
>>> from urllib import unquote, unquote_plus
>>> unquote("Hello+world%2Bfoo")
'Hello+world+foo'
>>> unquote_plus("Hello+world%2Bfoo")
'Hello world+foo'
```

**Comment by matitalatina** [comment]
I simply replaced `encoded = quote_plus(string, safe="~/")` to `encoded = quote(string, safe="~/")`  in `S3.py` line `985`. I'm using it with ceph and it works well.
This is my fork https://github.com/waveinch/s3cmd/blob/master/S3/S3.py
If you want to try it just:
`pip install git+git://github.com/waveinch/s3cmd.git`

**Comment by matitalatina** [comment]
@fviard thanks for the explanation. But unfortunately this is the only way to get things work with ceph... If I upload a file named `a a+a a.png` when I download it, I get `a+a+a.png`... With `quote` I have the correct behaviour.

**Comment by fviard** [comment]
@matitalatina : You can do as you want, but I would not advise you to do this that way. Because the good implementation of clients / servers will now expect it to be quoted_plus and so you could surprisingly be incompatible in the future. Or you could possibly encounter "signature does not match" with the backup of special filenames or so in unexpected cases.

My suggestion would be try to find in CEPH the thing that does the unquote instead of the unquote_plus to fix it. But take care, in a case I have seen like that, it was not the s3 service itself (like ceph) that was the issue, but a gateway or load balancer in front that was basically doing something like this:
Internet--> [ unquote (...) ; process; quote (...) ] ---> Ceph like service.

So, the thing in the middle looked innocent, but did destroy the paths as not based on "_plus" at least for the "unquote".

**Comment by jamshid** [comment]
IMO the suggested change to use quote() instead of quote_plus() is good.

I think it's more correct for s3cmd to send requests for objects "`hello world.html`" and "`C++.html"` like `GET /hello%20world.html` and `GET /C%2b%2b.html`, and I don't see how this change would break any AWS S3 users.

1. s3cmd is sending paths containing spaces encoded like `GET /hello+world.html` instead of `GET /hello%20world.html`.
2. AWS S3 accepts either format, will deliver the object named "`hello world.html`" (a space is in the key).
3. Technically, the first URI is *not* properly encoded. The "`+`" character is reserved  (https://tools.ietf.org/html/rfc3986#section-2.2) so it should be encoded.
  - A "`+`" (plus) character must be encoded as `%2b`, which would deliver an object whose key has a plus character.
  -  A "`+`" (plus) character should only be decoded to a "` `" (space) when used in *query args* (that rule is an HTML forms thing).
  - Since the Request-URI is not properly encoded I think AWS S3 could, maybe should, return a 400 error. Or, maybe behavior is undefined and it would have also been correct to treat it as a literal plus. Whatever, this "unconventional and unfortunate" behavior is here to stay: https://forums.aws.amazon.com/message.jspa?messageID=244218
4. s3cmd using `quote()` here instead of `quote_plus()` is more correct according to RFCs **and** it doesn't break AWS S3 users (AWS handles correctly encoded URIs just fine). 
5. A benefit of this (one line?) change would be that S3-compatible systems do not have to implement an incorrect encoding rule, at least for s3cmd. I don't think I've seen this problem with other S3 clients.

**Comment by fviard** [comment]
@jamshid Not true in my opinion.

- For the why quote_plus: 
The reason is that objects are url, and as of today, url are expected to be in quote_plus format:
http://stackoverflow.com/questions/1634271/url-encoding-the-space-character-or-20
 
- For the why s3cmd uses quote_plus:
AWS S3 expects quote_plus as object format, that is 100% sure.
Only using "quote" could work most of the times, but I'm pretty sure to have encountered a lot of issues with signature and so when "quote_plus" was not used.
Anyway, you can also have a look at the Internet:
https://forums.aws.amazon.com/thread.jspa?messageID=722673

- Even if the choice was open, most well done clients and servers are respecting aws interface that is using quote_plus, so not using that would mean that we could not work with a lot of non-aws servers as they are trying to mimic aws s3.

**Comment by jamshid** [comment]
See this answer on that stackoverflow question:
http://stackoverflow.com/a/29948396
> The reserved characters are different for each part.
> 
> For HTTP URLs, a space in a path fragment part has to be encoded to "%20" (**not, absolutely not "+"**), while the "+" character in the path fragment part can be left unencoded.
> 
> Now in the query part, spaces may be encoded to either "+" (for backwards compatibility: do not try to search for it in the URI standard) or "%20" while the "+" character (as a result of this ambiguity) has to be escaped to "%2B".

Yes, I understand that AWS S3 (somewhat incorrectly) treats "+" in a path like a space, but that's okay. We're all just saying that s3cmd does not **have to** rely on that. s3cmd can follow the URI spec precisely and AWS S3 will still work, and it will work with S3-compatible systems that do not implement AWS S3's  "unconventional and unfortunate" behavior.

Other S3 clients do **not** send "+" for space. I tried Cyberduck and verified with View => Toggle Log Drawer that a request to delete "hello world.html" is sent as:
`DELETE /mybucket/hello%20world.html HTTP/1.1`
I also tried `rclone --dump-headers -v copy 'hello world.html' s3:mybucket` and it shows
`PUT /mybucket/hello%20world.html HTTP/1.1`

**Comment by fviard** [comment]
@jamshid I just found an old thread on amazon forum with the reply to this question by an amazon representative:
https://forums.aws.amazon.com/thread.jspa?threadID=55746

If you would like them to be more compliant, it can probably be good for you to do a follow up on their forum to ask if they have no plan to change this behavior.
The point is that there could be some servers like ceph and some clients that don't use quote_plus, but a lot more others that are compliant with the Amazon protocol. So that can't be changed.
A test that you can do also, is to upload a file through amazon s3 official web interface and then use s3cmd in debug mode to see what was used with a "list" command.

**Comment by fviard** [comment]
@jamshid I was intrigued by the fact that aws now asks for a "%20" in the calculated headers for the signature. So I dug the internet.
Probably aws s3 is now resilient to quote or quote_plus being used.
Crazy as it is, there is not a single official page about this subject except the one stating that "keys should be url encoded".

So I had a look at boto and aws official java sdk, and it looks like that they did the switch to "quote" like behavior.
boto:
https://github.com/boto/boto/commit/0cd457906bc53d2f55d8a4b028bf8505b2fbbb17
aws-java-sdk:
https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3URI.java#L260

So in fact it could be good to do the switch to quote and so have a similar behavior as boto and aws-java-sdk.
But I'm wondering if there is not a risk to have a lot of third parties S3-like servers broken with such a change?

@mdomsch What do you think about that?

**Comment by fviard** [comment]
(reply to myself), in fact the encoding used by s3cmd before around 1 year ago was using "%20", so maybe changing could not have a too big impact.

**Comment by jamshid** [comment]
Seems like the suggested change was implemented in s3cmd 2.0, thanks!
I see s3cmd 2.0 now correctly sends requests with space encoded as "`%20`":
```
$ s3cmd --debug abortmp "s3://mybucket/mountainduck/CentOS 6 64-bit-000005.vmdk" cecdfd352e46148a09b288e1cb0cde33f83a0dbc011b1d8c76d99ef7b1e812a9f83a0dbc011b1d8c76d99ef7b1e812a90U
...
DEBUG: Sending request method_string='DELETE', uri=u'/mountainduck/CentOS%206%2064-bit-000005.vmdk?uploadId=cecdfd352e46148a09b288e1cb0cde33f83a0dbc011b1d8c76d99ef7b1e812a9f83a0dbc011b1d8c76d99ef7b1e812a90U'
```
whereas 1.6.1 would send this, which is not strictly correct (plus only means space in query args) and would break some S3-compatible servers.
```
DEBUG: Sending request method_string='DELETE', uri=u'/mountainduck/CentOS+6+64-bit-000005.vmdk?uploadId=cecdfd352e46148a09b288e1cb0cde33f83a0dbc011b1d8c76d99ef7b1e812a9f83a0dbc011b1d8c76d99ef7b1e812a90U'
```
So, this can probably be closed.
